### PR TITLE
Add credentials to SA for docker-registry exposed by a route

### PR DESCRIPTION
**Pull request by mfojtik**
If a docker-registry service is exposed by a route, this PR will make sure we generate the dockercfg automatically when that route is created or changed.

```
# run openshift with OPENSHIFT_DEFAULT_REGISTRY=registry.local:80
$ oc expose service docker-registry --hostname=registry.local --as=system:admin
```

Now when pushing in build:

```
Pushing image registry.local:80/test/ruby-ex:latest ...
...
Push successful
```

Fixes: https://github.com/openshift/origin/issues/8581

### CODE REVIEW

**Comment by mfojtik** [comment]
@deads2k PTAL

[test]

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 1]
```diff
@@ -248,6 +251,7 @@ func (e *DockerRegistryServiceController) getDockerRegistryLocations() []string
 		return []string{
 			net.JoinHostPort(service.Spec.ClusterIP, fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
 			net.JoinHostPort(fmt.Sprintf("%s.%s.svc", service.Name, service.Namespace), fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
+			net.JoinHostPort(fmt.Sprintf("%s.%s.svc.%s", service.Name, service.Namespace, defaultClusterDNSName), fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: This doesn't look right.  You can configure this right?

Comment (deads2k) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: Does this properly regenerate when the name changes?

Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: @deads2k how you can configure it? I mean even if you configure it for origin, will it work? (it is hardcoded in pkg/dns for example.

Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: @deads2k no

Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: see: https://github.com/openshift/origin/issues/8581#issuecomment-266692574

**Comment by mfojtik** [comment]
[test]

**Comment by wanghaoran1988** [comment]
Will this support what ever dns name we set with the `OPENSHIFT_DEFAULT_REGISTRY` ? like I want to use my special dns name different with the one that cluster used, eg, I like to use docker-registry.openshift.com.cn .

**Comment by mfojtik** [comment]
@wanghaoran1988 no, this will only support `docker-registry.default.svc.cluster.local`.

**Comment by mfojtik** [comment]
@deads2k what options do we have here? the desire is basically I want to have secrets for the registry I configured as my default registry (foo.bar.com:5000 - like I have route attached to the registry)

* check if there is a Route attached to "docker-registry" and use the route DNS (and regenerate when it changes)
* check the OPENSHIFT_DEFAULT_REGISTRY env var?

**Comment by deads2k** [comment]
> check if there is a Route attached to "docker-registry" and use the route DNS (and regenerate when it changes)
> check the OPENSHIFT_DEFAULT_REGISTRY env var?

It was structured with the intent to do both and dynamically regenerate whatever needed regenerating.

**Comment by mfojtik** [comment]
@deads2k so do we want to add both options? (check the route associated with the service and (re)generate dockercfg secret for it and also check the OPENSHIFT_DEFAULT_REGISTRY (if no route) and generate secret for it as well?) 

The second seems maybe overkill to me. I would assume that somebody expose the registry via Route to make it accessible publicly via public domain so we can just read the hostname from the Route.
IF you just set the env var and don't have route, then I would assume what you doing is setting `/etc/hosts` to point to registry clusterIP (which should not be what we recommend to users to do? ;-)

**Comment by deads2k** [comment]
> so do we want to add both options?

As long as they regenerate and you have add each, why not add both?

**Comment by deads2k** [comment]
needs tests.  I'm specifically interested in changing routes.

**Comment by mfojtik** [comment]
@deads2k will this test do? PTAL

File: pkg/serviceaccounts/controllers/docker_registry_service_test.go [hunk 2]
```diff
@@ -150,28 +171,72 @@ func TestUpdateNewStyleSecret(t *testing.T) {
...
+			switch {
+			case action.Matches("update", "secrets"):
+				updateService := action.(core.UpdateAction)
+				secret := updateService.GetObject().(*kapi.Secret)
+				actualDockercfg := &credentialprovider.DockerConfig{}
+				if err := json.Unmarshal(secret.Data[kapi.DockerConfigKey], actualDockercfg); err != nil {
+					t.Errorf("unexpected err %v", err)
+					continue
+				}
+				if !reflect.DeepEqual(*actualDockercfg, expectedDockercfgMap) {
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/docker_registry_service_test.go]: pass this as an arg.  I couldn't follow the code.

Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service_test.go]: sure

**Comment by deads2k** [comment]
> @deads2k will this test do? PTAL

minor comment, otherwise the test is fine.

**Comment by mfojtik** [comment]
@deads2k fixed, let the test pass

**Comment by mfojtik** [comment]
flake: https://github.com/openshift/origin/issues/12273

[test]

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 3]
```diff
@@ -242,16 +279,27 @@ func (e *DockerRegistryServiceController) getDockerRegistryLocations() []string
...
-		return []string{
+		result = append(result, []string{
 			net.JoinHostPort(service.Spec.ClusterIP, fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
 			net.JoinHostPort(fmt.Sprintf("%s.%s.svc", service.Name, service.Namespace), fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
+		}...)
+	}
+
+	if objs := e.routeCache.List(); len(objs) > 0 {
+		route := objs[0].(*routeapi.Route)
+		// TODO: Can a router have ports other than 80?
```
Comment (miminar) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: Yes, it can: https://github.com/openshift/origin/issues/11820

Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: ok that makes it tricky... how do I know which router is serving this route?

Comment (miminar) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: You could iterate over all of the routers and call the below on those matching route according to doc: https://docs.openshift.org/latest/install_config/router/default_haproxy_router.html#using-router-shards

I presume we already have some helper for router matching based on route labels.

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 4]
```diff
@@ -242,16 +279,27 @@ func (e *DockerRegistryServiceController) getDockerRegistryLocations() []string
...
+		if route.Spec.TLS == nil {
+			result = append(result, net.JoinHostPort(route.Spec.Host, "80"))
+		} else {
+			result = append(result, net.JoinHostPort(route.Spec.Host, "443"))
```
Comment (miminar) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: Also, default ports (`80` and `443`) shouldn't be included.

Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: @miminar i'm reworking this now to honor the OPENSHIFT_DEFAULT_REGISTRY

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 5]
```diff
@@ -242,16 +285,37 @@ func (e *DockerRegistryServiceController) getDockerRegistryLocations() []string
...
 	hasClusterIP := (len(service.Spec.ClusterIP) > 0) && (net.ParseIP(service.Spec.ClusterIP) != nil)
 	if hasClusterIP && len(service.Spec.Ports) > 0 {
-		return []string{
+		result = append(result, []string{
 			net.JoinHostPort(service.Spec.ClusterIP, fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
 			net.JoinHostPort(fmt.Sprintf("%s.%s.svc", service.Name, service.Namespace), fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
+		}...)
+	}
+
+	// Generate dockercfg entry for value of OPENSHIFT_DEFAULT_REGISTRY (if set)
```
Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: @miminar @deads2k this makes is a little bit more robust and deal with OPENSHIFT_DEFAULTS_REGISTRY variable (test updated as well).

The supported mechanisms are basically:

1) You expose the docker-registry via Route (`registry.foo.local`) -> we generate entry for `registry.foo.local` (with no ports, assuming docker will figure out whether it is 80 or 443).
2) You expose via Route but create Routers that routes into a non-standard port (eg. `registry.foo.local:5000`). In that case, you have to set a `OPENSHIFT_DEFAULT_REGISTRY=registry.foo.local:5000) as we can't tell what port the router is using.
3) You don't expose any route, just set `OPENSHIFT_DEFAULT_REGISTRY` - in that case we just create secret for the URL you provided (and regenerate on every master restart).

The 1) and 2) regenerate secrets everytime you change the Route hostname on live server. The 2) also allows to do crazy thing like set the `OPENSHIFT_DEFAULT_REGISTRY=docker-registry.default.svc.cluster.local` but also expose it via public route. Both will get the secrets generated properly.

Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: In most cases users will want to expose registry via route and set `OPENSHIFT_DEFAULT_REGISTRY` to that hostname so builds will use that registry for pushing. That should work well.

Users that don't want route, but still use named internal service can set this to `docker-registry.default.svc.cluster.local:5000` and it should work for them as well.

File: pkg/cmd/server/origin/run_components.go [hunk 6]
```diff
@@ -171,10 +172,12 @@ func (c *MasterConfig) RunServiceAccountPullSecretsControllers() {
 	dockerRegistryControllerOptions := serviceaccountcontrollers.DockerRegistryServiceControllerOptions{
 		RegistryNamespace:    "default",
 		RegistryServiceName:  "docker-registry",
+		RegistryDefaultHost:  os.Getenv("OPENSHIFT_DEFAULT_REGISTRY"),
 		DockercfgController:  dockercfgController,
 		DockerURLsIntialized: dockerURLsIntialized,
 	}
-	go serviceaccountcontrollers.NewDockerRegistryServiceController(c.KubeClientset(), dockerRegistryControllerOptions).Run(10, make(chan struct{}))
+	routeClient, _ := c.RouteAllocatorClients()
```
Comment (deads2k) [inline pkg/cmd/server/origin/run_components.go]: Go ahead and use the privileged loopback directly.  That client isn't for this.

File: pkg/serviceaccounts/controllers/docker_registry_service.go (hunk 5 above)
Comment (deads2k) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: > You expose the docker-registry via Route (registry.foo.local) -> we generate entry for registry.foo.local (with no ports, assuming docker will figure out whether it is 80 or 443).

Check the docs.  I'm not sure it will.  @pweil- don't I have to explicitly list my ports?

Comment (deads2k) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: Eh, actually, I must not in the config file, since I just looked in mine and it wasn't set.

**Comment by smarterclayton** [comment]
We are going to need to change the image API to provide both internal and external specs for clients, or do something really complicated with homing, but we cannot push by default to a route for lots of reasons, including scale.  Pull specs inside the cluster and pull specs outside the cluster are different in practice, even if it would be desirable to remove the need to distinguish between them.  We *are* going to need to make "external registry URL" a config argument to the image API in order to serve externally visible names, so let's not add any assumptions that contradict that.

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 7]
```diff
@@ -79,6 +87,33 @@ func NewDockerRegistryServiceController(cl kclientset.Interface, options DockerR
 	e.servicesSynced = e.serviceController.HasSynced
 	e.syncRegistryLocationHandler = e.syncRegistryLocationChange
 
+	e.routeCache, e.routeController = framework.NewInformer(
```
Comment (smarterclayton) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: This is cray cray.  External registry names are special, and we need a better mechanism to do so than looking for special routes.

**Comment by mfojtik** [comment]
@smarterclayton the context is basically users that set up a route for the integrated registry and then set the `OPENSHIFT_DEFAULT_REGISTRY` pointing to that route.

We document this as [exposing registry externally](https://docs.openshift.com/container-platform/3.3/install_config/registry/securing_and_exposing_registry.html#exposing-the-registry).

However, as a consequence of doing this, the builds are not able to authorize against that registry route since the docker secret is missing that route.

I agree that the registry and image API are linked and having an "external" pull spec on the Image API object makes sense to me (so as a user I can share the image via public URL outside the cluster). But that does not solve the problem of `OPENSHIFT_DEFAULT_REGISTRY` variable which if set, the builds are pushing to that registry without having proper credentials (and since it is there for 2 releases (?) I don't think we can just disable it).

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 8]
```diff
@@ -242,16 +285,37 @@ func (e *DockerRegistryServiceController) getDockerRegistryLocations() []string
...
+	}
+
+	// If the 'docker-registry' service is exposed by a Route, add the dockercfg entry for
+	// the Route hostname only if the OPENSHIFT_DEFAULT_REGISTRY is not already set to the
+	// same hostname.  This prevents from having duplicate entry with no port.
+	//
+	// Note that this allows to expose the 'docker-registry' only on the 80 or 443 ports.
+	// If you have router that expose registry on non-standard port, you should set the
+	// OPENSHIFT_DEFAULT_REGISTRY instead.
+	if objs := e.routeCache.List(); len(objs) > 0 {
```
Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: fixed.

**Comment by mfojtik** [comment]
@smarterclayton after conversation with @soltysh I think we don't need this PR at all. This PR will really just allow you to push from a build to an internal registry exposed by an external route. We should make it clear, that we are not going to support pushing to a "route" but always use the internal service IP address (for performance reasons).

Unless, you set a `OPENSHIFT_DEFAULT_REGISTRY` to an external registry (Artificatory, ...). In that case, we can't generate the security credentials for you and you will have to add your Artifactory push credentials manually to the SA.

Additionally, we agreed with @soltysh that we want to present public pull spec URL to users instead of presenting the image pull spec in web console and CLI using the service IP address (I will create Trello card for that). We should be able to automatically annotate/set field on the IS with external pull spec IF the registry is exposed by a public route.

**Comment by smarterclayton** [comment]
Couple of scenarios I can think of in the future:

1. I have a router that imposes a certain form of security (requires access
to a given domain).  The secret we generate wouldn't match that router no
matter what we do.

2. I have a router that is exposed to the external world and want my users
to use it (maybe this case), but only in certain cases

In the former, auto adding a secret won't break (since user can specify
multiple secrets for same registry).  In the latter, it also won't hurt.

So I do think we may eventually want to do this, but it may be via a config
value "imageRegistry.internalURL" (defaults to
registry.openshift.svc.cluster.local) and "imageRegistry.externalURL" (set
by installer to registry....).  If our config starts living in etcd, then
this change will be necessary, but I'd much prefer to do it based on config.

Also, duplicating large secrets for multiple urls is a problem.  We
probably can't afford to keep doing it if # of urls per secret increases.
Maybe we should ensure we can avoid having duplicate the secret contents.

On Jan 2, 2017, at 5:51 AM, Michal Fojtik <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> after conversation with
@soltysh <https://github.com/soltysh> I think we don't need this PR at all.
This PR will really just allow you to push from a build to an internal
registry exposed by an external route. We should make it clear, that we are
not going to support pushing to a "route" but always use the internal
service IP address (for performance reasons).

Unless, you set a OPENSHIFT_DEFAULT_REGISTRY to an external registry
(Artificatory, ...). In that case, we can't generate the security
credentials for you and you will have to add your Artifactory push
credentials manually to the SA.

Additionally, we agreed with @soltysh <https://github.com/soltysh> that we
want to present public pull spec URL to users instead of presenting the
image pull spec in web console and CLI using the service IP address (I will
create Trello card for that). We should be able to automatically
annotate/set field on the IS with external pull spec IF the registry is
exposed by a public route.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/12244#issuecomment-269957184>,
or mute
the thread

**Comment by sdodson** [comment]
@mfojtik bump!

**Comment by mfojtik** [comment]
@sdodson going to revive this tomorrow, sorry for the delay.

**Comment by brenton** [comment]
@mfojtik, what's the status on this?  Reading the history I see at one point we thought this PR wasn't needed.  Your comment from last week makes me think there's been more background discussions.  Ultimately, I'm trying to find out if there are any remaining blockers for https://bugzilla.redhat.com/show_bug.cgi?id=1242961 which I believe requires the ability to push to the internal registry using DNS.

cc @sdodson

**Comment by mfojtik** [comment]
@brenton @sdodson this should not be needed to support pushing to registry via DNS. We already generating the builder secrets with DNS records...

This PR was about generating secrets in case you set the `OPENSHIFT_DEFAULT_REGISTRY` to point to an internal registry exposed via a route. We agreed above that we are not going to address this use-case, but instead, we are going to expose the public pull spec via image stream (so as a user you will know how you can pull the image).

So to summarize:

1) We have trello card now to support public pull spec on image stream (should be addressed in 3.7/3.8)
2) We need to verify that we generating secrets for the "correct" registry.
3) We should switch the code which provides IP:PORT of the internal registry to return the DNS instead.

**Comment by brenton** [comment]
Thanks for the update @mfojtik.  For step 3, will your team open a PR for that?

**Comment by mfojtik** [comment]
@brenton just send email to Paul and Mike about this ;-)

We need to:

1) Update the generator to produce `.cluster.local`, assuming that is important (@sdodson)
2) Update https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/master_config.go#L209 to use DNS instead of IP address (@smarterclayton will that be OK?).

The 2) should force all image stream/builds/etc to use DNS over the IP address. The 1) will allow builds to push to registry.

**Comment by sdodson** [comment]
> Update the generator to produce .cluster.local, assuming that is important (@sdodson)

I'd prefer this but if this is too much trouble we can add the dns domain to the search path of the host's /etc/resolv.conf

> Update https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/master_config.go#L209 to use DNS instead of IP address (@smarterclayton will that be OK?).

We can also set OPENSHIFT_DEFAULT_REGISTRY='docker-registry.default.svc:5000' for the master to test with.

**Comment by brenton** [comment]
@sdodson, should I just ask QE to regression test by setting `OPENSHIFT_DEFAULT_REGISTRY='docker-registry.default.svc:5000'` and setting the /etc/resolv.conf search path to include `cluster.local`?  If this was a supported workaround for now I'm sure many customers would be happy.

**Comment by sdodson** [comment]
@brenton yeah that's fine, we do have a card to do this work in sprint 130 so we should implement it soon.

**Comment by abessifi** [comment]
@sdodson @brenton @mfojtik 

I'm using OpenShift v1.5.1 (deployed with Ansible and ran in containerized mode) and have all images tagged (pushed/pulled) with the service ip address instead of its DNS. As I understood, setting OPENSHIFT_DEFAULT_REGISTRY='docker-registry.default.svc:5000' fixes the 'issue' and forces OpenShift to contact the registry with its dns name.

Could you please point me out how to passe/activate the `OPENSHIFT_DEFAULT_REGISTRY` environement variable to/in OpenSHift ?

**Comment by sdodson** [comment]
@abessifi Assuming you've installed openshift using openshift-ansible you'll have dnsmasq configured for you. You can set `OPENSHIFT_DEFAULT_REGISTRY='docker-registry.default.svc:5000'` in /etc/sysconfig/origin-master, restart the master service and ensure that 'cluster.local' is in the search path in /etc/resolv.conf on all your hosts and it should work.

All of this is done for you on clean 3.6 installs, we don't migrate people during upgrade because we need to re-write the registry cert and that's a certificate that's commonly customized but we'll soon re-deploy the cert if it's a cert that hasn't been customized.

**Comment by abessifi** [comment]
@sdodson Thanks for your quicl reply.
Now I can see in the build logs "pushing to docker-registry.default.svc.cluster.local:5000 ..." but the push failed and crashed this error: `unauthorized: authentication required`.

It seems by default the serviceaccount builder/deployer/default have `dockercfg` secrets which didn't contain an authentication entry for "docker-registry.default.svc.cluster.local:5000" and editing these secrets didn't get the staff working (still get 'authentication required' error).

Here is the error messages I've got from the docker-registry pod output when trying to push my image after the build:

**Comment by sdodson** [comment]
Did you set `OPENSHIFT_DEFAULT_REGISTRY='docker-registry.default.svc:5000'` or `OPENSHIFT_DEFAULT_REGISTRY='docker-registry.default.svc.cluster.local:5000'` I don't think secrets are created for the latter in 1.5.

**Comment by abessifi** [comment]
When setting `OPENSHIFT_DEFAULT_REGISTRY='docker-registry.default.svc:5000'` I get an error telling no certificates for `docker-registry.default.svc`. And when I tryied with `OPENSHIFT_DEFAULT_REGISTRY='docker-registry.default.svc:5000'` I get the above authentication error instead.
